src/lambda_function/raw_sync/app.py

```diff
@@ -0,0 +1,518 @@
+            the basename without any extensions.
+        expected_key (str): The key of the corresponding raw object.
+        raw_keys (dict): A dictionary formatted as the dictionary returned by `list_s3_objects`.
+
+    Returns (str): The matching S3 key from `raw_keys`, or None if no match is found.
+    """
+    logger.debug(f"Expecting to find matching object at {expected_key}")
+
+    # Navigate through raw_keys to locate the correct `data_type` and `cohort`
+    if data_type in raw_keys:
+        if cohort in raw_keys[data_type]:
+            # Iterate through the list of keys under the specified `data_type`` and `cohort`
+            for key in raw_keys[data_type][cohort]:
+                if key == expected_key:
+                    logger.debug(f"Found matching object {expected_key}")
+                    return key
+    return None
+
+
+def parse_content_range(content_range: str) -> tuple[int, ...]:
+    """
+    Parse the ContentRange header to extract the start, end, and total size of the object.
+
+    A helper function for `list_files_in_archive`.
+
+    Args:
+        content_range (str): The ContentRange header value in the format 'bytes start-end/total'.
+
+    Returns:
+        tuple: A tuple containing (range_start, range_end, total_size).
+    """
+    # ContentRange format: 'bytes start-end/total'
+    _, range_info = content_range.split(" ")
+    range_start, range_end, total_size = map(
+        int, range_info.replace("-", "/").split("/")
+    )
+    logger.info(
+        f"Read 0-indexed bytes from {range_start} to {range_end}, inclusive, "
+        f"out of {total_size} bytes."
+    )
+    return range_start, range_end, total_size
+
+
+def unpack_eocd_fields(body: bytes, eocd_offset: int) -> list[int]:
+    """
+    Extract the End of Central Directory (EOCD) fields from the given body.
+
+    A helper function for `list_files_in_archive`.
+
+    The `unpack` method parses out:
+
+    <  - indicates Little-endian byte order
+    4s - 4-byte string:
+        EOCD signature (4 bytes)
+    4H - Four 2-byte unsigned short integers:
+        Number of this disk (2 bytes)
+        Disk where central directory starts (2 bytes)
+        Number of central directory records on this disk (2 bytes)
+        Total number of central directory records (2 bytes)
+    2L - Two 4-byte unsigned long integers
+        Size of central directory (4 bytes)
+        Offset of start of central directory (4 bytes)
+    H  - 2-byte unsigned short integer
+        Comment length (2 bytes)
+
+    Args:
+        body (bytes): The byte content from which to extract EOCD fields.
+        eocd_offset (int): The offset position of the EOCD signature in the body.
+
+    Returns:
+        tuple: A tuple containing (central_directory_offset, central_directory_size).
+            Both are int type.
+    """
+    eocd_fields = struct.unpack("<4s4H2LH", body[eocd_offset : eocd_offset + 22])
```

Comment:
Is it possible to get an invalid `eocd_offset` (not -1 but something else)? Do we have handling for that?

Reply:
Not according to [the docs](https://docs.python.org/3/library/stdtypes.html#bytes.rfind). See `determine_eocd_offset` for where that method is used.